hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/RowStatisticsImpl.java

```diff
@@ -0,0 +1,275 @@
+      tooLarge = true;
+    }
+    if (cellSize > blockSize) {
+      cellsLargerThanOneBlock++;
+      tooLarge = true;
+    }
+
+    if (tooLarge && LOG.isDebugEnabled()) {
+      LOG.debug("CellTooLarge: size={}, blockSize={}, maxCacheSize={}, table={}, cell={}", cellSize,
+        blockSize, maxCacheSize, table, CellUtil.toString(cell, false));
+    }
+
+    if (cellSize > largestCellBytes) {
+      largestCellRef = cell;
+      largestCellBytes = cellSize;
+    }
+    valueSizeBuckets.add(cell.getValueLength());
+
+    totalCells++;
+    if (CellUtil.isDelete(cell)) {
+      totalDeletes++;
+    }
+    totalBytes += cellSize;
+  }
+
+  /**
+   * Clone the cell refs so they can be cleaned up by {@link Shipper#shipped()}. Doing this lazily
+   * here, rather than eagerly in the above two methods can save us on some allocations. We might
+   * change the largestCell/largestRow multiple times between shipped() calls.
+   */
+  public void shipped(RawCellBuilder cellBuilder) {
+    if (largestRowRef != null) {
+      largestRow = CellUtil.cloneRow(largestRowRef);
+      largestRowRef = null;
+    }
+    if (largestCellRef != null) {
+      largestCell = RowStatisticsUtil.cloneWithoutValue(cellBuilder, largestCellRef);
+      largestCellRef = null;
+    }
+  }
+
+  public String getTable() {
+    return table;
+  }
+
+  public String getRegion() {
+    return region;
+  }
+
+  public String getColumnFamily() {
+    return columnFamily;
+  }
+
+  public boolean isMajor() {
+    return isMajor;
+  }
+
+  public byte[] getLargestRow() {
+    return largestRow;
+  }
+
+  public String getLargestRowAsString() {
+    return Bytes.toStringBinary(getLargestRow());
+  }
+
+  public long getLargestRowNumBytes() {
+    return largestRowBytes;
+  }
+
+  public int getLargestRowCellsCount() {
+    return largestRowCells;
+  }
+
+  public Cell getLargestCell() {
+    return largestCell;
+  }
+
+  public String getLargestCellAsString() {
+    return CellUtil.toString(getLargestCell(), false);
+  }
+
+  public long getLargestCellNumBytes() {
+    return largestCellBytes;
+  }
+
+  public int getCellsLargerThanOneBlockCount() {
+    return cellsLargerThanOneBlock;
+  }
+
+  public int getRowsLargerThanOneBlockCount() {
+    return rowsLargerThanOneBlock;
+  }
+
+  public int getCellsLargerThanMaxCacheSizeCount() {
+    return cellsLargerThanMaxCacheSize;
+  }
+
+  public int getTotalDeletesCount() {
+    return totalDeletes;
+  }
+
+  public int getTotalCellsCount() {
+    return totalCells;
+  }
+
+  public int getTotalRowsCount() {
+    return totalRows;
+  }
+
+  public long getTotalBytes() {
+    return totalBytes;
+  }
+
+  public Map<String, Long> getRowSizeBuckets() {
+    return rowSizeBuckets.toMap();
+  }
+
+  public Map<String, Long> getValueSizeBuckets() {
+    return valueSizeBuckets.toMap();
+  }
+
+  @Override
+  public String toString() {
+    return ("RowStatistics{" + "largestRow=" + Bytes.toStringBinary(largestRow)
```

Comment:
Instead of building these by hand, better to use commons-lang3 ToStringBuilder with its SHORT_PREFIX_STYLE.